ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs

```diff
@@ -498,10 +500,11 @@ forkBlockProduction maxBlockSizeOverride IS{..} BlockProduction{..} =
     -- another node was also elected leader and managed to produce a block
     -- before us, the header right before the one at the tip of the chain.
     prevPointAndBlockNo :: SlotNo
+                        -> BlockNo
                         -> AnchoredFragment (Header blk)
                         -> Either SlotNo (Point blk, BlockNo)
```

Comment:
Let's change this `BlockNo` into the block number of the block *we're going to produce*, not the block number of the predecessor, because *there might not be a predecessor*. The docstring needs to be updated and the name of this function too.

Reply:
Instead of returning the slot from the future as `Left`, return the actual error, since we can fail not only because of a slot from the future, but also when there's already a block in the slot we're trying to produce a block for, and, moreover, that block is already immutable. This might happen in two cases: 1) the clock moved backwards, on restart we ignored everything from the VolatileDB since it's all in the future, and now the tip of the ImmutableDB (= anchor point of our empty chain fragment) points to a block produced in the same slot we're trying to produce a block in; 2) k = 0 and we already adopted a block from another leader of the same slot.

So introduce an actual error type for these two cases (use the above as part of its docstring) and add a new trace message for the extra case.